src/NUnitFramework/framework/Constraints/Comparers/ComparisonState.cs

```diff
@@ -0,0 +1,62 @@
+using System.Collections.Generic;
+
+namespace NUnit.Framework.Constraints.Comparers
+{
+    internal struct ComparisonState
+    {
+        /// <summary>
+        /// Flag indicating whether or not this is the top level comparison.
+        /// </summary>
+        public readonly bool TopLevelComparison;
+
+        /// <summary>
+        /// A list of tracked comparisons
+        /// </summary>
+        private readonly Comparison[] _comparisons;
+
+        public ComparisonState(bool topLevelComparison)
+        {
+            TopLevelComparison = topLevelComparison;
+            _comparisons = new Comparison[0];
+        }
+
+        private ComparisonState(bool topLevelComparison, Comparison[] comparisons)
+        {
+            TopLevelComparison = topLevelComparison;
+            _comparisons = comparisons;
+        }
+
+        public ComparisonState PushComparison(bool topLevelComparison, object x, object y)
+        {
+            var comparisons = new Comparison[_comparisons.Length + 1];
+            _comparisons.CopyTo(comparisons, 0);
```

Comment:
what is your thinking of using an array and copying every time you add to it instead of a collection class like List?

Reply:
Here's the reason (with example) of why something like this is needed: https://github.com/nunit/nunit/pull/3406#discussion_r339322348.

It would be nice to avoid array copies, especially if there are ever graphs that go more than three or four levels deep. In order to not hold up the PR, I'd be happy to implement an ImmutableStack based on an singly-linked list instead of using array storage. Unless implementing a minimal ImmutableStack sounds interesting to you, @stevenaw :D

Reply:
Sorry, I must've misunderstood your initial suggestion when you mentioned the immutablestack. By all means go for it :) My free time over the coming free weeks is going to be a little more restricted, so I welcome any help or collaboration you want in this area.
Within this PR, I guess the main requirements would be ability to push, and ability to enumerate all items. Thanks @jnm2 !

Reply:
No worries, I tried to leave that open-ended. If you like, you can pull or cherry-pick this commit into your PR, or I can do a separate PR after this one. Up to you.

https://github.com/stevenaw/nunit/compare/issue52/cyclical-graph...nunit:examples/cyclical_graph_immutablestack

Reply:
Thanks! I've done a fast-forward merge of your branch into mine and run some tests. Pushed + just waiting for CI to finish